DOCS-12525 [EDITS]Copy to edit (#180)
mulesoft/docs-mule-sdk: modules/ROOT/pages/rest-sdk/edits/customization-layer.adoc

= Create a Customization Layer
ifndef::env-site,env-github[]
include::../_attributes.adoc[]
endif::[]

> After creating your connector, optionally create an instance of your existing

Reviewer: Do you create an instance or create another instance (of the original...instance? Is the existing one also an instance or is it just a placeholder/design thingie?
I guess I don't understand why you don't just created the customized connector that you want, rather than creating a descriptor, a connector, and then another layer for "further" customization (before you have actually customized the connector?), and then customizing...the original descriptor and connector? Eek, my brain hurts.

Reviewer: Other than my brain hurting, your text is super clean, so if you know that your audience understands, you are good to go except for the copyediting.

Author: The naming for these terms is a little poor actually, so your questions are valid! The customization layer --> customizing the connector have way too similar names even though they do different things. I'm hoping the info on the customization layer page clears up more of these questions but I'm adding more to the end of the previous section too.

Author: The create connector name is a little weird but basically:

you can either:
scaffold connector descriptor -> create connector -> customize the connector 

OR 

scaffold connector descriptor -> create connector -> add customization layer -> customize the connector 

you are technically "customizing" before you even "customize the connector". You technically "customize"  even in the scaffold connector descriptor steps -- "A connector descriptor is a YAML file that enables you to describe a connector, customize aspects of the generated connector (such as name, Maven GAV, operation names, layout information, and so on). think of the scaffold connector descriptor as one part in which you can customize but you can add another layer (customization layer) to describe other use cases through metadata. An example would be if you have a simple Anypoint connector, but if you want to add Composer functionalities BUT ALSO have keep the Anypoint version, you can use the customization layer to add these functionalities (so now it is two layers).

connector descriptor to generate a new layer of functionality and customization

> to describe other use cases through metadata.

Reviewer: So does this mean creating one layer per use case or one connector per use case?

== Customization Layer

> The customization layer is an instance of an existing connector descriptor.

Reviewer: Does this mean the same thing?
```suggestion
The customization layer is an instance of your existing connector descriptor that enables you to create a separately released, new version of your connector by making changes to your connector's existing code. This new version does not affect your original connector nor break backward compatibility with <what?>.
```

When you create a connector using the customization layer, the connector is
versioned and released separately from the base connector. Using the customization
layer enables you to develop new versions of the base connector without duplicating code.
Changes that are made in the new version do not affect the base version of the connector,

> nor does it break backward compatibility.

Reviewer: I got a little confused by "base" connector and felt like I wanted to make clear that this "base" is actually the connector that you just created using the connector descriptor you just created: yes? So this is just a simple way to create a customized version of the original that is separate from the original, versus replacing the original via customization?

Reviewer: I'd like to see the term "base connector" used in the Create a Connector topic, just so it maps to this one.

Reviewer: It would be SWELL to have a conceptual graphic that illustrates this some time: the scaffolding and layers made visual.

Author: "a simple way to create a customized version of the original that is separate from the original, versus replacing the original via customization?" --- Yes!

I like that idea, added the base connector term to the create connector topic.

Using the customization layer enables you to take an immutable version of a connector
and generate net new assets that add new layers of functionality. These newly
generated composite artifacts can have their own release lifecycle and be developed

> separately from the base connector.

Reviewer: "net new assets"? = results from new use cases?
"composite artifacts"? = versions of base connector?
This para sounds a bit jargony to me.

== Create the Customization Layer

> The customization layer is a composition feature that is a part of the standard
> connector creation flow. The only difference is that multiple connector
> descriptors can be specified in this `createConnector` command.

Reviewer: Does this mean:
```suggestion
You create a customization layer as part of the standard
connector creation flow by specifying more than one connector
descriptor in the `createConnector` command:
```
I wonder if we want to "create" the layer or "add" the layer?

To create the customization layer, run the `createConnector` MOJO:

----
mvn com.mulesoft.connectivity:rest-sdk-mojo:<PLUGIN_VERSION>:createConnector -DapiSpec=api.raml -Ddescriptor=connector-descriptor.yaml,customization-layer.yaml
----

> Specify the descriptors in the order you want to use the descriptors.

Reviewer: Does this really mean specify the descriptors in the order of their...progression...from base to version?

In this case, the `connector-descriptor.yaml` is applied first, and is then overwritten by
the `customization-layer.yaml`.

== Composing Descriptors

When composing standard connector descriptors, the properties declared in a
descriptor override those declared in previous (base) descriptors.

For example, if the connector name and description are set in the first (base) layer
and then the connector name is overwritten in a second descriptor layer, the
final generated connector has the name of the second layer and the description

> of the first layer.

Reviewer: ...and does this really mean that whatever descriptors are not overwritten in the command remain the same as in the original?
I guess if I am getting this right, then all the apparent wordiness is doing its job. I always look to make things shorter and simpler, but maybe that is inappropriate.

== What's Next

> Now that you've created the descriptors for your connector,

Reviewer: This intro is misleading, because this topic was not about creating descriptors except in the context of creating a customization layer for versioning the base connector, right?
One of the problems with a What's Next cycle is the optional tasks like this one.
